### PR TITLE
Bug/51 list only unique entries

### LINKED ISSUE #51: Listing only unique entries

### PR DESCRIPTION
**Pull request by ckunki**
fixes #51

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -216,6 +251,37 @@ Needs: impl, itest
...
+
+This affects especially the list of `Bucket` contents.
+
+Rationale:
+* A hierarchical representation of files and folders provides additional benefits:
+  * Hierarchies are a convenient and familiar concept to users.
+  * Hierarchies enable operations on multiple entries in a common scope, e.g. list, copy, or delete.
+
+To support the coexistence of files and folders with the same name, folders should be represented with a slash `/` as suffix. The list of contents of a folder may then contain the same entry twice:
+* one time as file (without suffix)
```
Comment (redcatbear) [inline doc/design.md:282]:
```suggestion
* once as file (without suffix)
```

File: doc/design.md [hunk 2]
```diff
@@ -216,6 +251,37 @@ Needs: impl, itest
...
+* one time as folder (with suffix)
```
Comment (redcatbear) [inline doc/design.md]:
```suggestion
* a second time as directory (with suffix)
```

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java [hunk 3]
```diff
@@ -72,10 +74,32 @@ private ReadOnlyBucket createBucket() {
...
+
+    @Test
+    // [utest->dsn~bucket-lists-files-with-common-prefix~1]
+    void listSubFolder() throws Exception {
+        simulateResponse(lines("folder/b1.txt", "b.txt", "folder/a1.txt", "a.txt"), 200);
+        final List<String> actual = createBucket().listContents("folder/");
+        assertThat(actual, equalTo(List.of("a1.txt", "b1.txt")));
+    }
+
+    @Test
```
Comment (redcatbear) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java:92]: A test for the `a` + `a/b.txt` case seems to be missing.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java (hunk 3 above)
Comment (ckunki) [reply src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java:92]: Just below 😃

File: doc/design.md [hunk 4]
```diff
@@ -65,17 +69,48 @@ Covers:
 Needs: impl, itest
 
 ### List of `Bucket` Contents
-
 `dsn~bucket-lists-its-contents~1`
```
[the remark this answers is not in the document]
Comment (ckunki) [reply doc/design.md:72]: OK, wilco.

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java (hunk 3 above)
Comment (redcatbear) [reply src/test/java/com/exasol/bucketfs/ReadEnabledBucketTest.java:92]: Damn. :-)

File: doc/design.md [hunk 5]
```diff
@@ -216,6 +251,37 @@ Needs: impl, itest
...
+* `a/b.txt` is interpreted as *folder* `a` containing file `b.txt`
+
+This affects especially the list of `Bucket` contents.
+
+Rationale:
+* A hierarchical representation of files and folders provides additional benefits:
+  * Hierarchies are a convenient and familiar concept to users.
+  * Hierarchies enable operations on multiple entries in a common scope, e.g. list, copy, or delete.
+
+To support the coexistence of files and folders with the same name, folders should be represented with a slash `/` as suffix. The list of contents of a folder may then contain the same entry twice:
```
[the remark this answers is not in the document]
Comment (ckunki) [reply doc/design.md]: Already replaced.

File: doc/design.md [hunk 6]
```diff
@@ -216,6 +251,37 @@ Needs: impl, itest
...
+The design decides to interpret `Bucket` to contain a hierarchy of entries. Each entry may either be a *file* or a *folder*. An entry is a *folder* if it has children, otherwise the entry is a *file*. An entry has children when its name contains the BucketFS separator `/`.
+
+Examples:
+* `a.txt` is a *file*
+* `a/b.txt` is interpreted as *folder* `a` containing file `b.txt`
+
+This affects especially the list of `Bucket` contents.
+
+Rationale:
+* A hierarchical representation of files and folders provides additional benefits:
```
[the remark this answers is not in the document]
Comment (ckunki) [reply doc/design.md]: Already replaced.

File: doc/design.md (hunk 2 above)
Comment (ckunki) [reply doc/design.md]: Fixed.

File: doc/design.md [hunk 7]
```diff
@@ -65,17 +69,48 @@ Covers:
...
+
+Covers:
+* `req~bucket-content-listing~1`
+
+Needs: impl, utest
+
+### Append Suffix to Folders
+`dsn~bucket-lists-folders-with-suffix~1`
+
+The list of contents of the bucket contains folders with a slash `/` as suffix.
```
[the remark this answers is not in the document]
Comment (ckunki) [reply doc/design.md]: fixed.

**Review by redcatbear — APPROVED** [review]

**Pull request merged**